cli/switch_bin/switch.go

```diff
@@ -0,0 +1,147 @@
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/apex/log"
+	"github.com/fatih/color"
+	"github.com/manifoldco/promptui"
+	"github.com/tarantool/tt/cli/config"
+	"github.com/tarantool/tt/cli/list"
+	"github.com/tarantool/tt/cli/search"
+	"github.com/tarantool/tt/cli/util"
+	"github.com/tarantool/tt/cli/version"
+)
+
+// SwitchCtx contains information for switch command.
+type SwitchCtx struct {
+	// ProgramName is a program name to install.
+	ProgramName string
+	// Version of the program to install.
+	Version string
+}
+
+const ansi = "[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;" +
+	"[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))"
+
+func CleanString(str string) string {
+	var re = regexp.MustCompile(ansi)
+	return re.ReplaceAllString(str, "")
+}
+
+// ChooseProgram shows a menu in terminal to choose program for switch.
+func ChooseProgram(supportedPrograms []string) (string, error) {
+	programSelect := promptui.Select{
+		Label:        "Select program",
+		Items:        supportedPrograms,
+		HideSelected: true,
+	}
+	_, program, err := programSelect.Run()
+
+	return program, err
+}
+
+// ChooseVersion shows a menu in terminal to choose version of program to switch to.
+func ChooseVersion(cliOpts *config.CliOpts, programName string) (string, error) {
+	binDirFilesList, err := os.ReadDir(cliOpts.Env.BinDir)
+
+	if len(binDirFilesList) == 0 || errors.Is(err, fs.ErrNotExist) {
+		return "", fmt.Errorf("there are no binaries installed in this environment of 'tt'")
+	} else if err != nil {
+		return "", fmt.Errorf("error reading directory %q: %s", cliOpts.Env.BinDir, err)
+	}
+	versions, err := list.ParseBinaries(binDirFilesList, programName, cliOpts.Env.BinDir)
+	if err != nil {
+		return "", err
+	}
+	if len(versions) == 0 {
+		return "", fmt.Errorf("there are no %s installed in this environment of 'tt'", programName)
+	}
+	var versionStr []string
+	for _, version := range versions {
+		if strings.Contains(version.Str, "[active]") {
+			versionStr = append(versionStr, util.Bold(color.GreenString(version.Str)))
+			continue
+		}
+		versionStr = append(versionStr, color.YellowString(version.Str))
+	}
+	versionSelect := promptui.Select{
+		Label:        "Select version",
+		Items:        versionStr,
+		HideSelected: true,
+	}
+	_, version, err := versionSelect.Run()
+	version = CleanString(version)
+	if strings.Contains(version, "[active]") {
+		version = strings.TrimSuffix(version, " [active]")
+	}
+
+	return version, err
+}
+
+// switchTt switches 'tt' program.
+func switchTt(binDir string, switchCtx SwitchCtx) error {
+	log.Infof("Switching to %s %s.", switchCtx.ProgramName, switchCtx.Version)
+	versionStr := search.ProgramTt + version.FsSeparator + switchCtx.Version
+	if util.IsRegularFile(filepath.Join(binDir, versionStr)) {
+		err := util.CreateSymlink(versionStr, filepath.Join(binDir, "tt"), true)
+		if err != nil {
+			return err
```

Comment:
pls, add a context to the error, Like "program version switch failed" or "failed to switch version: %s". Otherwise, I will see only "symlink creation error" error message, which does not provide any info about the purpose.

Reply:
added